File: LeetCode_Algorithm/[re]1680(Concatenation of Consecutive Binary Numbers)/1680.js
var concatenatedBinary = function(n) {
  // sol3
  // 2^53-1을 넘는 값을 처리하지 못함
  // ans = ""; 

  // for (let i=1; i < n+1; i++) {
  //   ans += i.toString(2)
  // }
  // return parseInt(ans, 2) % (10**9+7)

  // sol4
  binary = []
  for (let num=1; num < n+1; num++) {
    temp = []
    while (num > 1) {
      temp.push(num & 1)
      num = num >> 1
    }
    if (num == 1){
      temp.push(1)
    }
    temp.reverse()
    binary = binary.concat(temp)
  }
  ans = 0; power = 0
  for (let i=binary.length-1; i >= 0; i--) {
    if (binary[i] == 1) {
      ans += 2 ** power
      ans = ans % (2**10+7)
    }
    power += 1
  }
  return ans
};

console.log(concatenatedBinary(3))
console.log(concatenatedBinary(42))
n = 2

console.log(n >> 1)
console.log(n >> 1)

File: LeetCode_Algorithm/[re]1680(Concatenation of Consecutive Binary Numbers)/1680.py
class Solution:
    def concatenatedBinary(self, n: int) -> int:
        # sol1 time limit exceeded
        # total = ""
        # for num in range(1, n+1):
        #     temp = ""
        #     while num > 0:
        #         if num & 1:
        #             temp += '1'
        #         else:
        #             temp += '0'
        #         num = num >> 1
        #     total += "".join(list(reversed(list(temp))))
        # i = 1; ans = 0
        # for idx in range(len(total)-1, -1, -1):
        #     if total[idx] == '1':
        #         ans += i
        #     i *= 2
        # return ans % (10 ** 9 + 7)


        # sol2 time limit exceeded
        # 각 숫자마다 n의 시간 복잡도를 가지며 n개가 필요하므로
        # O(n^2) 
        # ans = 0; i = 1
        # for num in range(n, 0, -1):
        #     while num > 0:
        #         if num & 1:
        #             ans += i
        #         num = num >> 1
        #         i *= 2
        # return ans % (10 ** 9) + 7

        # sol3 2750ms 49%
        # O(nlogn) O(n)
        mod = (10 ** 9)+7
        ans = ""
        ans += "".join(bin(i)[2:] for i in range(n+1))
        return int(ans, 2) % mod

        # sol4 time limit exceeded
        # binary = []
        # for num in range(1, n+1):
        #     temp = []
        #     while num > 0:
        #         temp.append(num & 1)
        #         num = num >> 1
        #     binary.extend(list(reversed(temp)))
        # ans = 0; power = 0
        # for i in range(len(binary)-1, -1, -1):
        #     if binary[i] & 1:
        #        ans +=  2 ** power
        #        ans = ans % (10**9+7)
        #     power += 1
        # return ans
        


        
print(Solution().concatenatedBinary(3))
print(Solution().concatenatedBinary(12))

                    
